hazelcast/src/main/java/com/hazelcast/client/impl/management/ClientConnectionProcessListener.java

```diff
@@ -0,0 +1,212 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.client.impl.management;
+
+import com.hazelcast.cluster.Address;
+
+import java.util.EventListener;
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import static java.util.Arrays.asList;
+
+final class ListenerAggregate
+        implements ClientConnectionProcessListener {
+
+    private final List<ClientConnectionProcessListener> childListeners;
+
+    ListenerAggregate(ClientConnectionProcessListener... listeners) {
+        childListeners = new CopyOnWriteArrayList<>(asList(listeners));
+    }
+
+    @Override
+    public void attemptingToConnectToAddress(Address address) {
+        childListeners.forEach(listener -> listener.attemptingToConnectToAddress(address));
+    }
+
+    @Override
+    public void connectionAttemptFailed(Address target) {
+        childListeners.forEach(listener -> listener.connectionAttemptFailed(target));
+    }
+
+    @Override
+    public void hostNotFound(String host) {
+        childListeners.forEach(listener -> listener.hostNotFound(host));
+    }
+
+    @Override
+    public void possibleAddressesCollected(List<Address> addresses) {
+        childListeners.forEach(listener -> listener.possibleAddressesCollected(addresses));
+    }
+
+    @Override
+    public void authenticationSuccess(Address address) {
+        childListeners.forEach(listener -> listener.authenticationSuccess(address));
+    }
+
+    @Override
+    public void credentialsFailed(Address address) {
+        childListeners.forEach(listener -> listener.credentialsFailed(address));
+    }
+
+    @Override
+    public void clientNotAllowedInCluster(Address address) {
+        childListeners.forEach(listener -> listener.clientNotAllowedInCluster(address));
+    }
+
+    @Override
+    public void clusterConnectionFailed(String clusterName) {
+        childListeners.forEach(listener -> listener.clusterConnectionFailed(clusterName));
+    }
+
+    @Override
+    public void clusterConnectionSucceeded(String clusterName) {
+        childListeners.forEach(listener -> listener.clusterConnectionSucceeded(clusterName));
+    }
+
+    @Override
+    public void remoteClosedConnection(Address address) {
+        childListeners.forEach(listener -> listener.remoteClosedConnection(address));
+    }
+
+    @Override
+    public ClientConnectionProcessListener withAdditionalListener(ClientConnectionProcessListener listener) {
+        childListeners.add(listener);
+        return this;
+    }
+}
+
+/**
+ * Implementations can be attached to a {@link com.hazelcast.client.impl.connection.tcp.TcpClientConnectionManager} to receive
+ * fine-grained events about the client-to-cluster connection process.
+ *
+ * @since 5.2
+ */
+public interface ClientConnectionProcessListener
```

Comment:
what about methods `connectionAdded` and  `connectionRemoved` as in `ConnectionListener`?

Reply:
The [MC-side PR](https://github.com/hazelcast/management-center/pull/5799) doesn't need those currently.